for/schaefer.for

```diff
@@ -0,0 +1,216 @@
+
+      ! derivative of stress with respect to incremental plastic strain
+      ! because S = C: (eps_vec - Eoldp - dEp) and C, eps_vec, Eoldp are constants
+      ! youre left with the negative of the stiffness matrix
+      dS22ddEp = -C(2, :)
+      dS12ddEp = -C(ndir + 1, :) 
+
+      ! for convenience defined th (theta) to be f - b2 * S22. This term shows up a bunch
+      th = f- b2 * S22
+
+      ! Partial of f with respect to plastic strain. 
+      dfddEp = (S22 / th + b2) * dS22ddEp + ((a6 * S12) / th) * dS12ddEp
+
+      ! zero out J1
+      J1 = zero
+
+      ! J1 is a (ndir + nshr) x (ndir + nshr) matrix. It is the derivative of J0 with respect to dEpi.
+      ! A column "i" represents the terms corresponding to the dEpi derivative
+      DO i = 1, ndir + nshr, 1
+        ! fill in dSdEp
+        dSddEp = zero
+        dSddEp(2) = dS22ddEp(i)
+        dSddEp(ndir + 1) = dS12ddEp(i)
+
+        ! local derivatives
+        dS12 = dS12ddEp(i)
+        dS22 = dS22ddEp(i)
+        df = dfddEp(i)
+
+        ! get the the term corresponding to the ith partial derivative of dfdSdfdS
+        ! with respeoct to Epi
+        ! this derivative corresponds to the (2, 2) 
+        dfdSdfdSd22Ep = (b2 + S22 / th) * ( (2 * dS22) / th + (2 * (b2 * dS22 -df) * S22) / th**2) 
+        ! this derivative corresponds to the (2, 4) 
+        dfdSdfdSd24Ep = (a6 * dS12/th) * (b2 + S22/th) +&
+            (a6 * S12 / th**2) * (b2 + S22 / th) * (b2 * dS22 - df) +&
+            (a6 * S12 / th) * (dS22 / th + (b2 * dS22 - df)* S22 / th**2)
+        ! this derivative corresponds to the (4, 4) 
+        dfdSdfdSd44Ep =  (2 * a6**2 * S12 * dS12) / th**2 +&
+            (a6**2 * (2 * b2 * dS22 - 2 * df ) * S12**2) / th**3
+        ! Fill with all zeros and then fill its componenets as theyre calculated
+        dfdSdfdSdEp = zero
+        dfdSdfdSdEp(2, 2) = dfdSdfdSd22Ep
+        dfdSdfdSdEp(2, ndir + 1) = dfdSdfdSd24Ep
+        dfdSdfdSdEp(ndir + 1, 2) = dfdSdfdSd24Ep
+        dfdSdfdSdEp(ndir + 1, ndir + 1) = dfdSdfdSd44Ep
+
+        J1a = (n - 1) * f**(n - 2) * df * MATMUL(dfdSdfdS, dS)
+        J1b = f**(n - 1) * MATMUL(dfdSdfdSdEp, dS) 
+        J1c = f**(n - 1) * MATMUL(dfdSdfdS, dSddEp)
+        J1(:, i) = eye(:, i) - n * A * (J1a + J1b + J1c)
+      END DO
+      ! update according dEp according to newton raphson
+      IF (ndir + nshr .eq. 6) THEN
+        dEp = dEp - MATMUL(MInverse6x6(J1), J0)
+      ELSE IF (ndir + nshr .eq. 4) THEN
+        dEp = dEp - MATMUL(MInverse4x4(J1), J0)
+      ELSE 
+        Call log%error('Schaefer material prepeak nonlinearity model only supports ndir + nshr=6 and 4 elements')
+      END IF
+
+      ! The newton raphson continually updates J0 (and as it converges J0) should
+      ! approach 0. Determine its norm (and refer to this as the residual)
+      residual = SQRT(DOT_PRODUCT(J0, J0))
+      ! once we're within a certain tolerance break the loop
+      if (residual .lt. 1e-6) THEN
```

Comment:
For all convergence criteria and loop limits, please either define a named, commented variables in the subroutine, e.g.,
schaefer_nr_tolerance = 1d-6  ! convergence criterion for Newton-Raphson loop
counter_limit =1000000  ! maximum number of Newton-Raphson loops before divergence is assumed
...or create user-definable parameters and add them to the parameters module.